extensions/resteasy-classic/resteasy-common/deployment/src/main/java/io/quarkus/resteasy/common/deployment/ResteasyCommonProcessor.java

```diff
@@ -232,8 +232,13 @@ JaxrsProvidersToRegisterBuildItem setupProviders(BuildProducer<ReflectiveClassBu
                 index, beansIndex);
 
         if (useBuiltinProviders) {
+            contributedProviders.removeAll(availableProviders);
             providersToRegister = new HashSet<>(contributedProviders);
-            providersToRegister.addAll(availableProviders);
+            availableProviders.removeAll(providersToRegister);
+            availableProviders.removeAll(annotatedProviders);
+            for (String availableProvider : availableProviders) {
+                reflectiveClass.produce(ReflectiveClassBuildItem.builder(availableProvider).fields().build());
+            }
```

Comment:
I think this is extremely confusing. If we have to do crazy things like that, we need to categorize the providers properly and not rely on names like `contributedProviders` or `availableProviders`. There's no way to actually figure out what's going on when reading this code without going line by line and try to figure out what they  could contain.